### PR TITLE
Make sure users enter correct city when creating ads

### PR DESCRIPTION
**Pull request by vlado**
This is a follow up on https://github.com/vlado/earthquake-croatia/pull/102

@mmorava I tested your changes with production data locally and I had to make some changes. I think that it is almost ready for merge (see below for the only issue I've found). Notable changes that I made:

* Splitted migration to 2 files, import cities and counties, connect them with ads (will probably break with. your current db so rollback or recreate the db).
* Made sure migration works with production data.
* Added new city (Zagreb) and New county (Inozemstvo) to be able to store foreign cities that are already used in ads.
* Cities filter now only has cities that are used in ads
* Tried to switch cities to grouped select but it has rendering issue. Maybe you can know what is wrong. If it is too hard I suggest we use standard select.

### CODE REVIEW

**Review by mmorava — APPROVED** [review]

File: app/controllers/ads_controller.rb [hunk 1]
```diff
@@ -5,9 +5,9 @@ class AdsController < ApplicationController
 
   # rubocop:disable Metrics/AbcSize
   def index
-    @ads = Ad.where(kind: ad_kind).order(created_at: :desc).paginate(page: params[:page], per_page: 20)
+    @ads = Ad.strict_loading.includes(:city).where(kind: ad_kind).order(created_at: :desc).paginate(page: params[:page], per_page: 20)
```
Comment (vfonic) [inline app/controllers/ads_controller.rb:8]: Oh yeah! Strict loading! 🤘

Comment (vlado) [reply app/controllers/ads_controller.rb:8]: 🔥

**Comment by vlado** [comment]
I've copied production database to review app on https://potres-features-filter--ovc3gb.herokuapp.com. It seems to me that data is ok but not sure why the select2 plugin is not working (no JS errors)?

@mmorava @vfonic @radanskoric @shime @berislavbabic @Xenosb @Macxim Any ideas what could be wrong?

File: db/migrate/20210103181832_connect_ad_with_city.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    end
+  end
+
+  def import_hr_ad(ad)
+    city_name =CITIES_MAPPING[ad.city_zdel] || ad.city_zdel
+    city = City.find_by(name: city_name) || City.find_by(zip_code: ad.zip)
+    if city
+      ad.update!(city_id: city.id )
+    else
+      raise "Can not foundi HR city `#{city_name}`"
```
[the remark this answers is not in the document]
Comment (vfonic) [reply db/migrate/20210103181832_connect_ad_with_city.rb:127]: This still fails on the first missing city in the db.

File: app/assets/stylesheets/cities.scss [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the cities controller here.
```
Comment (vfonic) [inline app/assets/stylesheets/cities.scss:1]: Let's get rid of these empty legacy scaffold files?

File: app/helpers/ads_helper.rb [hunk 4]
```diff
@@ -30,6 +30,10 @@ def kinds_for_select
     Ad.kinds.keys.map { |key| [t("ad.kinds.#{key}"), key] }
   end
 
+  def cities_for_filter_select
+    City.where(id: Ad.select("DISTINCT(city_id)")).order(:name).pluck(:name, :id)
```
Comment (vfonic) [inline app/helpers/ads_helper.rb:34]: I played around a bit and found that this query does the same thing, but in one db query as opposed to two:
```rb
City.joins(:ads).all.pluck(:name, :id).uniq
```

For this to work, you need to add `has_many :ads` to City model.

File: app/helpers/cities_helper.rb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module CitiesHelper
```
Comment (vfonic) [inline app/helpers/cities_helper.rb:3]: I'd get rid of this file as well

File: db/migrate/20210103181821_seed_locations.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+class SeedLocations < ActiveRecord::Migration[6.1]
+  def up
+    Rake::Task['db:seed_locations'].invoke
+  end
+
+  def down
+    City.delete_all
+    County.delete_all
```
Comment (vfonic) [inline db/migrate/20210103181821_seed_locations.rb:8]: Since we're doing so in other db migrations, we could do the same here (prevent migrations failing when/if we remove these models in the codebase):

```rb
  City = Class.new(ActiveRecord::Base)
  County = Class.new(ActiveRecord::Base)
```

**Review by vfonic — CHANGES_REQUESTED** [review]
This is still failing on the first city it cannot find in the db (in my case it failed so many times on `"Krizevci"`. Perhaps we should add it to the `cities_mapping.yml`?

We should also get rid of `NormalizeCityName` that's no longer in use as well as `cities_mapping.yml` and `ads.rake` which calls `cities_mapping.yml` and will no longer be relevant after this PR.

File: app/models/ad.rb [hunk 7]
```diff
@@ -55,14 +58,14 @@ class Ad < ApplicationRecord
 
   def to_param
     hr_category = I18n.t("ad.categories.#{category}")
-    [id, hr_category.parameterize, city.parameterize].join("-")
+    [id, hr_category.parameterize, city.name].join("-")
```
[the remark this answers is not in the document]
Comment (vfonic) [reply app/models/ad.rb:61]:
```suggestion
    [id, hr_category.parameterize, city.name.parameterize].join("-")
```

[the remark this answers is not in the document]
Comment (vlado) [reply app/models/ad.rb:61]: Nice catch 👍

[the remark this answers is not in the document]
Comment (vlado) [reply app/models/ad.rb:61]: Nice catch 👍

File: app/helpers/ads_helper.rb (hunk 4 above)
Comment (vlado) [reply app/helpers/ads_helper.rb:34]: `City.where(id: Ad.select("DISTINCT(city_id)")).order(:name)` also generates on query (with subquery) and does not return duplicates (no need for uniq). Haven't tested which one is faster :)

File: db/migrate/20210103181821_seed_locations.rb (hunk 6 above)
Comment (vlado) [reply db/migrate/20210103181821_seed_locations.rb:8]: 👍 Forgot to add them when writing down method.

**Comment by vlado** [comment]
Thanks for the review @vfonic. select2 works find for you locally also?

**Comment by vfonic** [comment]
@vlado 
The filtering works, but I found a couple of issues:
* [ ] select2 added its own dropdown arrow (and it's the only one that's working when dropdown is clicked
   the other dropdown uses Bulma's arrow, the two don't look alike :/
* [ ] select2 and Bulma's dropdown padding don't match. Bulma's: `padding-left: calc(.75em - 1px);` (this is 11px). select2's padding-left is 8px
* [ ] There's some delay when clicking on dropdown to open (almost 1sec delay)
   Can't we populate the list of cities in the dropdown on page load (no ajax calls at all)? We're only filtering based on cities that have at least one ad. I'd guess there's not a lot of those cities (so we can load it with the HTML perhaps?)
* [ ] we lost the "current filters" badges and I don't see a way to clear the filters
   That "Odustani" button could be repurposed for this.
* [ ] This one's tricky: The dropdown is showing all the cities with at least one ad, but we're always filtering by kind.
   If I go to "Tražim pomoć", there will be a huge amount of cities that will return no results when selected

**Comment by vfonic** [comment]
@vlado I know what's happening. select2's styles are not being loaded (any styles imported in JS files are not being loaded in fact).

Long story short (https://github.com/rails/webpacker/issues/2059#issuecomment-518748208):

> You probably have `extract_css: true` in your `webpacker.yml`. This removes `import '../scss/index';` and generates the equivalent stylesheet. You do need to add `stylesheet_pack_tag` yourself.

You need to add `stylesheet_pack_tag` next to `stylesheet_link_tag` (we're using both).

I think this should work (never tried it with turbolinks, so I'm not sure whether we need the data attribute or not):
```erb
    <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
    <%= stylesheet_pack_tag 'application', media: 'all' %>
    <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
```

**Comment by vlado** [comment]
> * [ ]  There's some delay when clicking on dropdown to open (almost 1sec delay)

Strange. I do not have delay. You have delay on filter dropdown or on ad form? 

>   Can't we populate the list of cities in the dropdown on page load (no ajax calls at all)? We're only filtering based on cities that have at least one ad. I'd guess there's not a lot of those cities (so we can load it with the HTML perhaps?)

City filter is populated only with cities used in the ads and is not using Ajax. Ajax is used in the ad form cause there are 6000+ cities and I was getting "JS Is blocking this page" error.

> * [ ]  we lost the "current filters" badges and I don't see a way to clear the filters

Fixed in https://github.com/vlado/earthquake-croatia/pull/107/commits/4de4b4a8d57e3a8498c62c252293495cccfd068f

> * [ ]  This one's tricky: The dropdown is showing all the cities with at least one ad, but we're always filtering by kind.
>   If I go to "Tražim pomoć", there will be a huge amount of cities that will return no results when selected

Fixed in https://github.com/vlado/earthquake-croatia/pull/107/commits/7c828a5a9a0865a413d16df52a538a0d56cf78fe

**Comment by vfonic** [comment]
> Strange. I do not have delay. You have delay on filter dropdown or on ad form?

It seems to only be happening when devtools are open. I managed to reproduce the issue in Chrome and Brave.

https://user-images.githubusercontent.com/67437/103749854-55890100-5006-11eb-9985-79d9992626d9.mov



There's no delay in Safari and in Firefox the dropdown closes as soon as it opens.
Firefox:

https://user-images.githubusercontent.com/67437/103749667-1490ec80-5006-11eb-931d-f9eca1b3518f.mov


Maybe some of these issues will go away when we have select2 styles added?
The Chrome delay issue, we could ignore it. It only happens when devtools are open.

**Comment by vlado** [comment]
> Maybe some of these issues will go away when we have select2 styles added?

Since I'm not that skilled in CSS, can someone try to fix the styles?
@vfonic @Macxim @Xenosb

**Comment by vfonic** [comment]
@vlado I'll have a look now

**Comment by vfonic** [comment]
@vlado this should be ready for review: https://potres-features-filter--ovc3gb.herokuapp.com/

Here's the list of changes I've made:
* Hide select2 dropdown arrow
  I figured that's the easiest. When dropdown opens, the arrown down stays as arrow down unfortunately. But it was too much work to change it to arrow up as the arrow element comes from Bulma and "is-expanded" is set on an element deeper in the DOM, so the only way to make this work would be to create a stimulus controller to handle this, so I chose not to do all that. :) 
* Fix checking if city filter is applied
  We were still checking `params[:city]` while it should be `params[:city_id]`
* Added `stylesheet_pack_tag` to prod environments
* Match select2 styles with Bulma dropdown styles
* Make sure select2 uses HR language translations
* New ad: Make sure users can type unaccented city names in search
  This uncovered this new issue:
  Screen Shot 2021-01-06 at 12 23 02
* New ad: Prevent showing "No results" if user just opened city search box

**Review by vfonic — APPROVED** [review]

**Comment by vfonic** [comment]
@vlado please have a look. I believe this is ready to be merged.

**Comment by vfonic** [comment]
@vlado I take that back. :) One sec. I'm fixing some issue.

**Comment by vfonic** [comment]
@vlado all good. We should be able to deploy this to live site now.

**Comment by vlado** [comment]
@vfonic looks perfect

I'll deploy later tonight.

* Put site to maintenance mode
* Import production database locally
* Run migrations locally
* Add mappings for newly added cities
* Deploy to production when all is ok locally
* Switch off maintenance mode (if needed after deploy)

**Comment by vlado** [comment]
@mmorava @vfonic This is not deployed and working fine on production. Thanks!
